pyccl/halos/concentration.py

```diff
@@ -0,0 +1,308 @@
+                                          np.atleast_1d(M),
+                                          a, mdef_other)
+
+        c = self.concentration(cosmo, M_use, a)
+        if np.isscalar(M):
+            c = c[0]
+        return c
+
+
+class ConcentrationDiemer15(Concentration):
+    name = 'Diemer15'
+
+    def __init__(self, mdef=None):
+        super(ConcentrationDiemer15, self).__init__(mdef)
+
+    def _default_mdef(self):
+        self.mdef = MassDef(200, 'critical')
+
+    def _setup(self):
+        self.kappa = 1.0
+        self.phi_0 = 6.58
+        self.phi_1 = 1.27
+        self.eta_0 = 7.28
+        self.eta_1 = 1.56
+        self.alpha = 1.08
+        self.beta = 1.77
+
+    def _check_mdef(self, mdef):
+        if (np.fabs(mdef.Delta - 200.) > 1E-4) and \
+           (mdef.rho_type != 'critical'):
+            return True
+        return False
+
+    def concentration(self, cosmo, M, a):
+        M_use = np.atleast_1d(M)
+
+        # Compute power spectrum slope
+        R = mass2radius_lagrangian(cosmo, M_use)
+        lk_R = np.log(2.0 * np.pi / R * self.kappa)
+        # Using central finite differences
+        lk_hi = lk_R + 0.005
+        lk_lo = lk_R - 0.005
+        dlpk = np.log(linear_matter_power(cosmo, np.exp(lk_hi), a) /
+                      linear_matter_power(cosmo, np.exp(lk_lo), a))
+        dlk = lk_hi - lk_lo
+        n = dlpk / dlk
+
+        sig = sigmaM(cosmo, M_use, a)
+        delta_c = 1.68647
+        nu = delta_c / sig
+
+        floor = self.phi_0 + n * self.phi_1
+        nu0 = self.eta_0 + n * self.eta_1
+        c = 0.5 * floor * ((nu0 / nu)**self.alpha +
+                           (nu / nu0)**self.beta)
+        if np.isscalar(M):
+            c = c[0]
+
+        return c
+
+
+class ConcentrationBhattacharya13(Concentration):
+    name = 'Bhattacharya13'
+
+    def __init__(self, mdef=None):
+        super(ConcentrationBhattacharya13, self).__init__(mdef)
+
+    def _default_mdef(self):
+        self.mdef = MassDef(200, 'critical')
+
+    def _check_mdef(self, mdef):
+        if mdef.Delta != 'vir':
+            if np.fabs(mdef.Delta - 200.) > 1E-4:
+                return True
+        return False
+
+    def _setup(self):
+        if self.mdef.Delta == 'vir':
+            self.A = 7.7
+            self.B = 0.9
+            self.C = -0.29
+        else:  # Now Delta has to be 200
+            if self.mdef.rho_type == 'matter':
+                self.A = 9.0
+                self.B = 1.15
+                self.C = -0.29
+            else:  # Now rho_type has to be critical
+                self.A = 5.9
+                self.B = 0.54
+                self.C = -0.35
+
+    def concentration(self, cosmo, M, a):
+        gz = growth_factor(cosmo, a)
+        status = 0
+        delta_c, status = lib.dc_NakamuraSuto(cosmo.cosmo, a, status)
+        sig = sigmaM(cosmo, M, a)
+        nu = delta_c / sig
+        return self.A * gz**self.B * nu**self.C
```

Comment:
no test for scalars here?

Reply:
Humm, I don't think we need that here. All functions of `M` here operate on scalars or arrays equally.

Reply:
Ahhh right.